### DISCUSSION TITLE
analogReadDiff questions

### DISCUSSION THREAD

**Discussion by dpetican** [SpenceKonde/DxCore#144] [Q&A]
I just tried doing a differential reading and am confused about something.  I don't seem to be getting an ADC reading that corresponds to the voltage of the signal.  My math/assumptions etc could be wrong.  This is the code I'm using to do the read:

`adc = analogReadDiff(17, 16, 12, 0);`

And the conversion to volts:

`#define ADC_VREF 3.33
#define ADC_SCALE (ADC_VREF / 4096.0 / 2.0)
volts = adc * ADC_SCALE;`

Again I don't understand why line breaks are not working in code blocks.  Sorry.

Also, I am using an external voltage reference hence 3.33V and a sample duration of 255 because sensor is very high impedance (100-100K resistance range). Hmmm should that be * 2.0 instead?  Thx.

**Comment by dpetican** [comment]
Yes the correct formula is:

`#define ADC_VREF 3.33
#define ADC_SCALE (ADC_VREF / 4096.0 * 2.0)
volts = adc * ADC_SCALE;`

Because its +/- 2048 counts.

**Comment by SpenceKonde** [comment]
Sorry, I didnt see this initially.

hmm. This functionality may not have been tested much on DxCore (the enhanced analogRead() functions were mostly developed on the 2-series to take advantage of it;s fancy ADC, but then hastily ported to DxCore and I'm not sure whether I did much testing, so there may be an issue here, I will investigate...

So, you have the voltage reference on one pin, and... what on the other? 

This is on a 28 or 32 pin part I take it? (if it;'s got more pins, you're reading from non-ADC pins)
Depending on the pincount, what pins 16 and 17 actually refer to will vary (pins are numbered sequentially counterclockwise, like the port pins are, but the port layout skips pins, while the core doesn't skip numbers of the pins. I always encourage people to use the PIN_Pxn constants - that way, no matter which pincount you compile for, it ends up on the same pins. 

At times I wonder if I should have declared that all pin numbers shall be (port * 8)+bit, so 28 pin part pin numbers would start at 0, go up to 7 (PORTA), then there's no PORTB so next is PORTC with pins 16, have 17, 18, and 19 - it's only got a half port there so then it would jump to 24 for PORTD, continue through to 31, etc - which would make the pin numbers and PIN_Pxn constants equally portable, and digitalPinToPort(pin) could just be (pin >> 3) and digitalPinToBitPosition(pin) would be (pin & 0x07), and a whole bunch of garbage would be optimized away). It would definitely have been wierder to work with though;

**Comment by dpetican** [comment]
Its the 28-pin part.  I got it working.  The rest about the sensor impedance and Vref don't matter much at this stage.  Overall, it seems to work well.  Just finishing up a hardware revision.  Will have more to say after I get the board back and write more code that uses the ADC.

As for pin numbering I was reluctant (old dog etc) to use the PIN_Pxx notation but now I like it.  Just drop the old pin numbers all together.  Is that possible/practical?

**Reply by SpenceKonde** [reply]
There's always internal numbering scheme of some sort. I guess the advantage to it being fully hidden is that then I could pick the pin numbers the way I wanted to (namelty, low 3 bits would store bit position, and the top 3 the port. 

I could get the port structure by bitwise and with 0xE0, cast to uint16_t add 0x400 the high byte of that pointer. All of the table lookups would vanish except for the TCBs. I think Im going to keep deemphasizing the old numbers, but I think it would inflict so muxh breakage on existing code that it would involve a new major version and declaration of incompatibility. The past few week;s I've seen some counterexamples to my predictions that the performance improvement would be large, so I'm less inclined to push on this as it stands. What I think I most regret though was skipping the UPDI pin.;..... Because if not for that, I would have had that by default on the 64-pin parts, and could have used that as a playground for optimization, see how much it buys and how worthwhile it is. the magnitude of the impact (also, because all those tables are actually in RAM not flash because these things have so much RAM nobody has noticed. I might at some point make "rationalized" pin mappings, but any approach is a ton of work - the compiler also just absolutely hates efficiently combining two bytes to make a uint16_t that you can cast to a pointer and use.  it frequently isn't smart enough t realize the that you can just copy the high byte to the register after the low byte and there's your value.   I've seen that a few times lately when I changed  something seeing performance enhancements and got very little for it, and that project has that risk too thee sort of change that I can see taking an assload of work and not delivering the performance improvements I'm hoping for. Im going to stick to focusing on changesthat are desperately needed urgently.

**Comment by SpenceKonde** [comment]
One thing to remember about the ADC in the Dx-series for differential measurements. The voltages you are measuring must both be less than VREF. which is a mondo bummer when the EA-series finally gets here it will have the 12-bit differential ADC with programmable gain amplifier like the 2-series has (and which prompted me to write that function!)
